### DISCUSSION TITLE
【更新日志】2024/04/14 PaoPaoDNS 重命名force_cn_list/force_nocn_list为force_recurse_list/force_dnscrypt_list

### DISCUSSION THREAD

**Discussion by kkkgo** [kkkgo/PaoPaoDNS#122] [更新日志]
- 增加debug调试信息和test测试项目、构建测试
- 改进特定情况下硬件资源计算策略避免在某些平台无法正常运行
- 改进在特定情况下无有效解析记录的时候指定返回码nxdomain

以上为更新日志补充，镜像日期大于2024/03/29均已包含以上更新。

## 重命名force_cn_list/force_nocn_list为force_recurse_list/force_dnscrypt_list 
因原有的`force_cn_list`/`force_nocn_list`会在很多情况下被望文生义造成用途上的误解，因此：
- `force_cn_list.txt`→将会重命名为`force_recurse_list.txt`
- `force_nocn_list.txt`→将会重命名为`force_dnscrypt_list.txt`
- 如果`/data`存在`force_cn_list.txt`或`force_nocn_list.txt`，也会被同时读取，但容器不会默认生成这两个文件在`/data`目录。**_因此升级容器镜像后不需要修改任何东西也不会影响你现有的配置。_**

## 关于list的用途，请参考新文档节选：

- `force_forward_list.txt`： 仅在配置`CUSTOM_FORWARD`有效值时生效，强制转发到`CUSTOM_FORWARD`DNS服务器的域名列表，容器版本更新不会覆盖该文件。一行一条，语法规则如下：  
以`domain:`开头域匹配: `domain:03k.org`会匹配自身`03k.org`，以及其子域名`www.03k.org`, `blog.03k.org`等。   
以`full:`开头，完整匹配，`full:03k.org` 只会匹配自身。完整匹配优先级更高。     
以`regexp:`开头，正则匹配，如`regexp:.+\.03k\.org$`。[Go标准正则](https://github.com/google/re2/wiki/Syntax)。   
以`keyword:`开头匹配域名关键字，如以`keyword: 03k.org`会匹配到`www.03k.org.cn`   
尽量避免使用`regexp/keyword`会消耗更多资源。域名表达式省略前缀则为`domain:`。同一文本内匹配优先级：`full > domain > regexp > keyword`     
- `force_dnscrypt_list.txt`：强制使用dnscrypt加密查询结果的域名列表，匹配规则同上。容器版本更新不会覆盖该文件。   
- `force_recurse_list.txt`：强制使用本地递归服务器查询的域名列表，*一般不会用到该list，强制递归的域名不会被生效CNFALL功能*，匹配规则同上。容器版本更新不会覆盖该文件。   
- `force_ttl_rules.txt`: 参见`RULES_TTL`功能。修改将实时重载生效。容器版本更新不会覆盖该文件。   
- 修改`force_forward_list.txt`或`force_dnscrypt_list.txt`或`force_recurse_list.txt`或`force_ttl_rules.txt`将会实时重载生效。
- 文本匹配优先级`(custom_mod功能seq: top)`>`force_forward_list` > `force_dnscrypt_list` > `force_recurse_list` > `force_ttl_rules`>`(custom_mod功能seq: list)`>`其他自动分流逻辑`。
- **注意事项**：由于跨平台系统差异，不建议使用Windows自带记事本编辑。如果list出现了问题无法读取或者无法生效，可以直接删除list文件，重启容器会自动重建默认的list。如果你想解析的域名位于境外，并且没有境内CDN，而你又想获取原始记录（与`force_forward_list.txt`或者使用`AUTO_FORWARD`功能获取到的解析记录区分开），那么你应该把域名加进`force_dnscrypt_list.txt`而不是`force_recurse_list.txt`，因为基于个人网络环境差异，递归服务器位于境外的域名存在递归失败的可能。*`force_recurse_list.txt`的应用场景一般应仅限于特殊域名递归调试，大部分场景都不适用于`force_recurse_list.txt`。* 此外，你可以根据`文本匹配优先级`灵活设置同一个域名子域名走不同的list。

**Comment by TwoBall** [comment]
请问是否可以实现pikpak主页和下载分开解析。类似https://community.nssurge.com/d/2444-pikpak

**Reply by TwoBall** [reply]
regexp:^dl-a10b-.+\.mypikpak\.com$
好像可以搞定

**Reply by kkkgo** [reply]
不需要正则，直接用full:就可以了。

**Reply by kkkgo** [reply]
配合ppgw就是这样：
`force_forward_list.txt`添加：
```yaml
full:mypikpak.com
domain:access.mypikpak.com
domain:access.mypikpak.net
domain:api-drive.mypikpak.com
domain:config.mypikpak.com
domain:download.mypikpak.com
domain:info.mypikpak.com
domain:rcv5-sp.mypikpak.com
domain:res-etl-ssl.mypikpak.com
domain:sg-thumbnail-drive.mypikpak.com
domain:static.mypikpak.com
domain:user.mypikpak.com
```
`force_dnscrypt_list.txt`添加：
```json
domain:mypikpak.com
```

**Reply by TwoBall** [reply]
不是，我的目的是让dl-a10b-0860.mypikpak.com这种不走代理，默认的mypikpak.com整个主域都是走代理的

**Reply by kkkgo** [reply]
我明白你的意思，但这是一个XY问题。你的目的是下载不走代理省流量，实际上mypikpak的服务器都是新加坡阿里云，直连都比你挂代理快，说白了把主域名和检测接口代理就能用还能省流量，哪天他改规则是s1-a10b-0860.mypikpak.com呢？按需代理显然就更加一劳永逸。
实际上需要真正需要代理的域名只有两个，添加到`force_forward_list.txt`:
```shell
full:mypikpak.com
full:access.mypikpak.com
```
`force_dnscrypt_list.txt`添加：
```json
domain:mypikpak.com
```
这样就应该能直接用了。

**Reply by TwoBall** [reply]
明白了，我开始没有理解两个规则的优先级。感谢🙏🏻

**Comment by 742038744** [comment]
大佬好，用了半年多了，一直有2个问题解决不了，我是爱快 + paopaodns + paopaogetway，psopaodns的ipv6是设置为yes，但是我的群晖会有ipv6使用的问题。一是群晖用一段时间后，ipv6默认网关就自动没了，需要手动关闭ipv6再重新打开来获取；第二是群晖获取ipv6地址和网关都正常的情况下，我用DDNS-go做了动态解析，用手机流量可以通过域名访问群晖，用家里路由器的有线或者无线都不行。

**Reply by kkkgo** [reply]
1、你用什么默认网关那是你路由器DHCP的事情，跟DNS无关。
2、参考文档`RULES_TTL`功能。或者参考 https://github.com/kkkgo/PaoPaoDNS/discussions/95
有其他和更新无关的问题可以单独开贴。

**Comment by LetMeDecay** [comment]
一个奇怪的问题，旧版我在cn里加的域名，会默认走路由器的解析，主机映射然后绑定到内网上
新版把域名加到force_recurse_list.txt里之后就默认走dns服务商那里的ip了，好像并没有经过路由器

**Reply by kkkgo** [reply]
`force_cn_list/force_recurse_list`是强制走递归，如果你和其他规则文件无重叠的话，递归成功了他的解析结果应该和真实托管在权威DNS上的解析结果一样。”路由器的解析“和”dns服务商“是指什么？

**Reply by TwoBall** [reply]
检查下你的dns是不是有本地网关(路由器lan ip)。

**Reply by TwoBall** [reply]
https://github.com/kkkgo/PaoPaoDNS/discussions/78#discussioncomment-7738200

**Reply by LetMeDecay** [reply]
路由器是openwrt，我配置了主机映射，把域名直接映射到内网了
在跑paopaodns的虚拟机上nslookup域名，返回的是内网地址，但是指定paopaodns来nslookup就会走真实的dns解析

**Reply by kkkgo** [reply]
你在openwrt设置主机映射那是dnsmasq的行为，你预期的解析结果也只会由dnsmasq提供的dns服务得到，你这设置跟paopaodns没有任何关系，跟其他dns服务也没有关系，除非你的openwrt存在dns劫持把其他dns服务的结果都篡改了。
你在虚拟机上使用由openwrt提供的dns服务自然可以得到你预期的结果，指定paopaodns来nslookup要是得不到真实的dns解析那才是有问题，那就成了被DNS劫持了。
如果你想在内网改写DNS记录，你可以参考`USE_HOSTS`功能或者`RULES_TTL`功能。或者参考https://github.com/kkkgo/PaoPaoDNS/discussions/95

**Reply by LetMeDecay** [reply]
用use_hosts解决了，奇怪的是之前是怎么解决的，难道是openwrt重定向把53端口劫持了

**Reply by kkkgo** [reply]
你可以跑test.sh检测有没有劫持。或者参考https://github.com/kkkgo/PaoPaoDNS/issues/2#issuecomment-1504708367

**Comment by oneone1995** [comment]
作者你好，已经将域名写到force_dnscrypt_list中，但获取到的仍然是fakeip。尝试删除list文件，重启容器后仍旧如此。测试了自动生成的几个域名，其中几个也是一样的情况。
```
# dig gh-proxy.com
; <<>> DiG 9.16.28 <<>> gh-proxy.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 15671
;; flags: qr aa rd ra ad; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 2

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; MBZ: 0x0001, udp: 4096
; COOKIE: ecbc72225f8809df (echoed)
;; QUESTION SECTION:
;gh-proxy.com.                  IN      A

;; ANSWER SECTION:
gh-proxy.com.           1       IN      A       11.0.15.44

;; ADDITIONAL SECTION:
20241006161706.249661604.addinfo.paopaodns. 2 IN TXT "Since 0ms From:forward -> 192.168.88.185@53"

;; Query time: 1 msec
;; SERVER: 192.168.88.206#53(192.168.88.206)
;; WHEN: Sun Oct 06 16:17:06 CST 2024
;; MSG SIZE  rcvd: 179

# dig gh-proxy.com +trace
; <<>> DiG 9.16.28 <<>> gh-proxy.com +trace
;; global options: +cmd
.                       580     IN      NS      g.root-servers.net.
.                       580     IN      NS      b.root-servers.net.
.                       580     IN      NS      i.root-servers.net.
.                       580     IN      NS      a.root-servers.net.
.                       580     IN      NS      j.root-servers.net.
.                       580     IN      NS      h.root-servers.net.
.                       580     IN      NS      d.root-servers.net.
.                       580     IN      NS      k.root-servers.net.
.                       580     IN      NS      l.root-servers.net.
.                       580     IN      NS      c.root-servers.net.
.                       580     IN      NS      m.root-servers.net.
.                       580     IN      NS      f.root-servers.net.
.                       580     IN      NS      e.root-servers.net.
.                       580     IN      RRSIG   NS 8 0 518400 20241019050000 20241006040000 61050 . RF/OYmvyYjZO8zFY1EKgXZjt8QK+pw/MLukcD8Im63DXxUW5+3/9ncpM +rO5cywXP6CpzOac9tg8tHOVrPOgqBg5t+7RkG9KxSUfYybXoB3L782c yL3+bi3ol0NOr8uEdufOEkwYkaJxbim7q8p0Q3WS9p8OPXxgX5J5eHxP eAUEpEtR5qJ+10gJ6Rf1juMY8x3Ni4GTF7wAY1oQq4J1QaxKQkOuemgE Rz4GZTgI9cRPBXr/ajR0Q6aatNlAMPnS90DYyIFk/sgS7ZxKYPRofpGm jyydvL+T55hDr74XBc/D93lJ9zgJuUTrRJXTdltszfHGBtHh/v5833NX O/A6cQ==
;; Received 1854 bytes from 192.168.88.206#53(192.168.88.206) in 0 ms
```

CNAUTO、AUTO_FORWARD、USE_MARK_DATA都为true

**Reply by kkkgo** [reply]
没有true的设置项，只有yes和no。建议单独开贴跑debug.sh

**Reply by oneone1995** [reply]
写帖子的时候写错了。。已经重新开贴啦
